src/video.cpp

```diff
@@ -791,24 +791,46 @@ namespace video {
     }
     display_wp = disp;
 
-    std::vector<std::shared_ptr<platf::img_t>> imgs(12);
-    auto round_robin = round_robin_util::make_round_robin<std::shared_ptr<platf::img_t>>(std::begin(imgs), std::end(imgs));
-
-    for (auto &img : imgs) {
-      img = disp->alloc_img();
-      if (!img) {
-        BOOST_LOG(error) << "Couldn't initialize an image"sv;
-        return;
+    constexpr auto capture_buffer_normal_size = 1;
+    constexpr auto capture_buffer_size = 12;
+    std::list<std::shared_ptr<platf::img_t>> imgs(capture_buffer_size);
+
+    auto pull_free_image_callback = [&](std::shared_ptr<platf::img_t> &img_out) -> bool {
+      img_out.reset();
+      for (auto it = imgs.begin(); it != imgs.end(); it++) {
+        // pick first unallocated or unused
+        if (!*it || it->use_count() == 1) {
+          if (!*it) {
+            // allocate if unallocated
+            *it = disp->alloc_img();
+          }
+          img_out = *it;
+          if (it != imgs.begin()) {
+            // move freshly allocated or unused img to the front of the list to prioritize its reusal
+            imgs.erase(it);
+            imgs.push_front(img_out);
+          }
+          break;
+        }
       }
-    }
+      if (img_out) {
+        // unallocate unused above normal buffer size
+        size_t index = 0;
+        for (auto &img : imgs) {
+          if (index >= capture_buffer_normal_size && img && img.use_count() == 1) img.reset();
```

Comment:
I think we should have a grace period before freeing the excess buffers to avoid thrashing alloc/free. I wrote logic like that in #1128 that you could use as a reference.

Reply:
Yes, it would make sense. But I think it's better not to shove too much logic into single big pull request that already contains potentially breaking changes. Multiple thresholds with different timeouts, simple to implement because the code is localized in one space. Just not everything at the same time, I'm struggling with macOS code portion at the moment, for example. And that is blocking the rest of no matter how perfect logic we write.

Reply:
Another example, we can save another frame of memory and do intelligent copy evasion for situations when there's no mouse cursor to blend. But that logic would be rather complex, so I decided to exclude it from this pull request.

Reply:
Adding grace period should be as simply as this https://github.com/ns6089/Sunshine/commit/0c95cc93e1b34146275892c12f294c84c1e64042 (untested, but logic is simple enough)
But I'd still rather not include it with this particular pull request, unless you insist.

Reply:
That's fine I guess. In your testing are you seeing it stick pretty consistently at 1 image or do you see it go up and down?

Reply:
Goes up to 2 at the very beginning (may indicate error in sleep scheduling there by the way), then stays solidly at 1. But that's nvidia, I'm half expecting amd having problems there as well. That's partially why I've added log messages with extensive statistics, need to gather baseline for tuning auto-resize. But I don't think it should hurt having very slight most likely invisible overhead for a couple of days in nightly branch.

Reply:
Either way, I've raised normal buffer size to 2 frames https://github.com/LizardByte/Sunshine/pull/1098/commits/7fa2c4f385152fd1daae03766a291bc568c5bc2d. Won't get downsized below this target. Nothing should block the merge now other than testing on Linux and macOS platforms. Temporary solution anyway, dynamic resize will be done here https://github.com/LizardByte/Sunshine/pull/1149.

Reply:
A safe choice, but in reality if it can't maintain one image and has to use 2, then we'd be safer going with the other pull requests that do the same thing with less changes. I am really hoping it can be brought down to 1, but if 2 is the minimum, it's still better than now I guess.

Reply:
I had thought the same thing initially, but I think this one has one very nice advantage over my simpler approach.

Because the new image is only retreived at the moment we're ready to capture into it, it means we can achieve an image pool of a single entry. Because the current code (and my #1128), latch the next image at the moment the previous one is submitted for encoding, it is limited to a 2 image minimum. There's basically no way we could manage to have just one image buffer without blocking the capture thread, since we need to grab the next one immediately after submitting the previous one.